scripts/eslint/rules/mark-jsdoc-types-used.cjs

```diff
@@ -0,0 +1,102 @@
+const { TSESTree, ESLintUtils } = require("@typescript-eslint/utils");
+const { createRule } = require("./utils.cjs");
+const ts = require("typescript");
+
+/**
+ * @param {ts.Identifier} node
+ */
+function isBareIdentifier(node) {
+    if (ts.isPropertyAccessExpression(node.parent)) {
+        return node.parent.expression === node;
+    }
+
+    if (ts.isQualifiedName(node.parent)) {
+        return node.parent.left === node;
+    }
+
+    // TODO(jakebailey): surely I'm missing something here.
+    // Also, how does eslint deal with type space versus value space? Or does it not?

```

Comment:
It does definitely - which is why we forked `eslint-scope`! Was a big piece of work.

ESLint core has no concept of it - it just knows values. Our forked scope manager tracks it and exposes the state via the `.isTypeVariable` and `.isValueVariable` properties on the `Variable`.

Using `context.markVariableAsUsed` eslint will just find the first variable matching the name and mark it as used (because it has no concept of type vs value). TBH this probably fine for this usecase - I doubt you guys have too much scope shadowing - but if you want you can pretty easily reimplement it with awareness (it's just a few lines).

https://github.com/eslint/eslint/blob/9e9edf93ecfa0658e8b79e71bc98530ade150081/lib/source-code/source-code.js#L690-L724

The weird thing about jsdoc is that it doesn't really care if it's a type or value for some cases, right?

Reply:
Is your forked one usable somewhere?

I know that what I've done is definitely going to be wrong because I'm pretty sure one can write a function type that introduces a new name and now it should refer to that one, but eslint's not going to see it.

I don't think JSDoc is any different in terms of type/value, though.

I know that this PR is also wrong in other ways, e.g. it probably marks parameters as used. There's some call I should be making that checks if a node is in type space or not but I forgot it before pushing my WIP change.

Reply:
> Is your forked one usable somewhere?

You get it for free by using our parser :)

————

> I know that what I've done is definitely going to be wrong because I'm pretty sure one can write a function type that introduces a new name and now it should refer to that one, but eslint's not going to see it.

¿que? Can you show me an example?

---

> I don't think JSDoc is any different in terms of type/value, though.

Yeah I meant in terms of "does an identifier in a JSDoc comment refer to a type or a value". Most places are unambiguous and can only reference one!

Reply:
> ¿que? Can you show me an example?

I mean that my method is naive, so given:

```ts
/** @typedef {number} T */

/** @type {{ <T>(x: T) => T }} */
const identity = x => x;
```

This is going to mark `T` as used because it's referenced in the param/return statement. So without someone managing the scopes, it'll mark that outer typedef as used. And given the API I'm using (where I'm just giving it names and no position info), I don't see how I can do it properly.

Reply:
It won't actually do that because JSDoc typedefs don't create variables in the scope analyser.

Our scope analyser only understands concrete TS syntax.
We don't support this because there's no sane reason to use a JSDoc typedef within a TS file.

If you want JSDoc typedefs marked as unused then yeah you'll need the JSDoc plugin (it does this I think?). You can scope the plugin to just JS files for that usecase to reduce the impact of it (unless there's some reason you guys use typedef within TS files..?)

Reply:
Yeah. I realized earlier that I should really be skipping most of these tags in TS files.

I'm mainly trying to avoid the JSDoc plugin because it's so slow, so I'll have to see if the scope manager is in any way available for use, or something else.

Reply:
Yeah so the problem is that comment nodes aren't attached to the AST - so you can't acquire the scope for it (`context.getSourceCode().getScope(node)`).

If you can figure out the node that contains the comment then you can acquire the scope and go from there based on the `markVariableAsUsed` code I linked above.
This isn't super easy to do though, sadly, because it involves traversing the AST to try and find the tightest containing node.

So I guess it really depends on how many violations there are in the codebase and what the violations look like exactly.

If you guys have test fixtures for your intellisense services or `noUnusedLocals` option then I'd just turn off the `no-unused-vars` rule on those files. (TBH generally you would turn off linting on test fixtures anyways).

Reply:
But you can see what I was saying about why we don't support this in our rule - it's a super complicated problem to solve on top of it being a niche ask!

As soon as josh mentioned it to me I was like "oh god no" heh.